### ISSUE TITLE
synchronization costs lots of time when multiple processes are doing it at the same time

### ISSUE THREAD

**Issue by Sunnydreamrain** [mila-udem/platoon#71]
Hi all,

I have noticed that the synchronization costs lots of time when multiple processes are doing it at the same time. Usually one synchronization costs about 2-4 secs. But it could be over 100 seconds when over 1 processes are doing it. Then all the processes are all waiting there (only one process is training). 

I tried to avoid  multiple processes synchronizing its parameters at the same time using the following code. Basically when one worker wants to synchronize its parameter, it will send a message to the server and check whether there is another worker doing it. However, it does not work. I have checked the code a few times and I think it should be able to solve the problem. Any idea why it does not work as expected? Thanks a lot.

worker

        status=worker.send_req('check_sync')
        while status['sync']==False:
          time.sleep(2)
          status=worker.send_req('check_sync')
        worker.sync_params(synchronous=True)
        redundunt=worker.send_req('reduce_sync')

controller

        elif 'check_sync' == req:
            if self.syncing:
                control_response=dict(sync=False)
            else: 
                control_response=dict(sync=True)
                self.syncing=True
        elif 'reduce_sync'==req:
            self.syncing=False   

In summary, two issues,
1, Why synchronization costs lots of time when multiple processes are doing it at the same time. (Probably waiting for the lock, but it takes much more time than needed.)
2, How to make sure only one worker synchronize its parameters at one time?

Thanks a lot.

**Comment by nouiz** [comment]
I don't have time to check the code. I don't know if other people will have
the time.

But we just merge a new way to synchronize that could make a big
difference. They all sync at the same time in a deterministic manner. Can
you check the new doc and try it? It work for GPU. I forgot if it work on
CPU.

**Comment by Sunnydreamrain** [comment]
Okay. Thanks.
